docs/architecture.md

```diff
@@ -0,0 +1,292 @@
+expectations.
+
+### Driver and Executor (and Cluster Manager) Network Communications
+
+One major requirement that **must** be met is network communication between
+Drivers and Executors. This allows Drivers to submit tasks to the Executors
+and collate their results. `armada-spark` achieves this through the use of
+[kubernetes labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/).
+Used correctly, these labels induce the Armada scheulder to schedule Drivers
+and their accompanying Executors to the same kubernetes cluster managed by
+Armada. Kubernetes cluster networking topology guarantees that nodes within the
+cluster will be able to communicate. Scheduling to the same cluster also has
+the benefit of increasing the probability of compute resources being
+"clustered" closely together. Perhaps servers on the same rack, or even a single
+server.
+
+If you refer back to the ["Cluster Manager Overview"](https://spark.apache.org/docs/latest/cluster-overview.html),
+the first image you'll find is a diagram showing a complete, directed graph between the:
+
+- Driver Program
+- Worker Node / Executors
+- Cluster Manager
+
+and while the diagram appears to separate the "Cluster Manager" out as a separate
+program or node, in reality it tends to "live" in the same process as the
+Driver. So by ensuring connectivity between the Driver and Executors, we also
+get connectivity with the "Cluster Manager".
+
+### Driver and Executors as Armada Jobs
+
+Armada, being a multi-cluster Kubernetes manager, takes on the role of
+scheduling "jobs" to the clusters it oversees. For now, it makes the most sense
+to separate each Driver and Executor into their own Armada job. There are 
+several considerations that must be observed when submitting these jobs.
+
+#### Armada Gang Scheduling
+
+`armada-spark` uses gang scheduling annotations to schedule the driver and 
+its executors together. By using the gang scheduling “Node Uniformity Label” 
+we can guarantee that a driver and its executors are scheduled to the same
+cluster if Kubernetes node labels are applied correctly.
+
+#### Kubernetes Node Labels
+
+Any Kubernetes worker nodes should have a common label, such as "armada-spark", 
+and each cluster should have a unique value assigned to that label.
+
+#### Custom Service Names
+
+A unique service name is assigned to each Spark job’s driver when submitting 
+a Spark job to Armada. This service name acts as a DNS name within Kubernetes 
+and allows Spark Executors to connect to their corresponding Drivers.
+
+### Armada Job Limitations
+
+```
+// TODO: Fill out
+```
+
+# `armada-spark` Deployment Considerations
+
+## Necessary Kubernetes Configuration
+
+### Kubernetes Cluster Nodes
+
+All nodes in a particular Kubernetes cluster must be labeled by a common
+`(label, value)` tuple. The `label` must be an agreed upon name which shows
+the cluster will accept `armada-spark` jobs. Something like `armada-spark`.
```

Comment:
I'd change
  accept `armada-spark` jobs. Something like `armada-spark`.
to
  accept `armada-spark` jobs, e.g. literally `armada-spark`.

Reply:
resolved in 8862a54